packages/analytics-core/src/utils/debug.ts

```diff
@@ -0,0 +1,90 @@
+/* eslint-disable @typescript-eslint/no-unsafe-assignment */
+/* eslint-disable @typescript-eslint/no-unsafe-member-access */
+/* eslint-disable @typescript-eslint/no-unsafe-call */
+import { Config, LogLevel, LogConfig } from '@amplitude/analytics-types';
+import { AmplitudeCore } from '../core-client';
+
+export const getStacktrace = (ignoreDepth = 0): string => {
+  const trace = new Error().stack || '';
+  return trace
+    .split('\n')
+    .slice(2 + ignoreDepth)
+    .map((text) => text.trim())
+    .join('\n');
+};
+
+// This hook makes sure we always get the latest logger and logLevel.
+export const getClientLogConfig = (client: AmplitudeCore<Config>) => (): LogConfig => {
+  const { loggerProvider: logger, logLevel } = { ...client.config };
+  return {
+    logger,
+    logLevel,
+  };
+};
+
+// This is a convenient function to get the attribute from object with string path, similar to lodash '#get'.
+export const getValueByStringPath = (obj: any, path: string): any => {
+  path = path.replace(/\[(\w+)\]/g, '.$1'); // convert indexes to properties
+  path = path.replace(/^\./, ''); // strip a leading dot
+  for (const attr of path.split('.')) {
+    if (attr in obj) {
+      obj = obj[attr];
+    } else {
+      return;
+    }
+  }
+  return obj;
+};
+
+export const getClientStates = (client: AmplitudeCore<Config>, paths: Array<string>) => (): { [key: string]: any } => {
+  const res: { [key: string]: any } = {};
+  for (const path of paths) {
+    res[path] = getValueByStringPath(client, path);
+  }
+  return res;
+};
+
+export const debugWrapper =
+  <T extends Array<any>, R>(
+    fn: (...args: T) => R,
+    fnName: string,
+    getLogConfig: () => LogConfig,
+    getStates?: () => { [key: string]: any },
+    fnContext: any = null,
+  ) =>
+  (...args: T): R => {
+    const { logger, logLevel } = getLogConfig();
+    // return early if possible to reduce overhead
+    if ((logLevel && logLevel < LogLevel.Debug) || !logLevel || !logger) {
+      return fn.apply(fnContext, args);
+    }
+    const logs: string[] = [];
+    logs.push(`start time: ${new Date().toISOString()}`);
+    logs.push(`function name "${fnName}" called with args: ${JSON.stringify(args, null, 2)}`);
+    if (getStates) {
+      logs.push(`states before execution ${JSON.stringify(getStates(), null, 2)}`);
+    }
+    logs.push(`function stacktrace: ${getStacktrace(1)}`);
```

Comment:
I am thinking of pushing objects instead. For example
```js
logs.push({
  startTime: new Date().toISOString(),
  fnName: fnName,
  fnArgs: JSON.stringify(args, null, 2),
  ...
})
```
Pushing string is not easy to change format.

Reply:
Thanks! Updated in [82139c2](https://github.com/amplitude/Amplitude-TypeScript/pull/254/commits/82139c2c12c8c673f7b55ef3eea652982bc5eb8e)!